### PR TITLE
feat: Add support for opt-in git deploy keys

### PR DESCRIPTION
**Pull request by pritamstyz4ever**
## Context

Screwdriver needs to support user configuration for using deploy keys for `git checkout`

## Objective

This PR allows users to opt-in for using git deploy keys and creates them on pipeline creation /update

## License
I confirm that this contribution is made under a BSD license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: test/plugins/pipelines.test.js [hunk 1]
```diff
@@ -1715,6 +1715,64 @@ describe('pipeline plugin test', () => {
             });
         });
 
+        it('returns 201 and correct pipeline dat with deployKey when annotations set', () => {
```
Comment (tkyi) [inline test/plugins/pipelines.test.js:1718]:
```suggestion
        it('returns 201 and correct pipeline data with deployKey when annotations set', () => {
```

File: plugins/pipelines/update.js [hunk 2]
```diff
@@ -6,6 +6,7 @@ const schema = require('screwdriver-data-schema');
 const idSchema = schema.models.pipeline.base.extract('id');
 const { formatCheckoutUrl, sanitizeRootDir } = require('./helper');
 const { getUserPermissions } = require('../helper');
+const ANNOTATION_USE_DEPLOY_KEY = 'screwdriver.cd/useDeployKey';
```
Comment (tkyi) [inline plugins/pipelines/update.js:9]: Wondering if it would be better to put these in a common file somewhere or leverage data-schema
but we can do that in a separate PR

File: plugins/pipelines/update.js [hunk 3]
```diff
@@ -136,6 +141,32 @@ module.exports = () => ({
...
+
+            if (deployKeyAnnotation) {
+                const deploySecret = await secretFactory.get({
+                    pipelineId: oldPipeline.id,
+                    name: deployKeySecret,
+                })
+
+                if (!deploySecret) {
+                    const privateDeployKey = await pipelineFactory.scm.addDeployKey({
+                        scmContext : oldPipeline.scmContext,
```
Comment (tkyi) [inline plugins/pipelines/update.js:155]:
```suggestion
                        scmContext: oldPipeline.scmContext,
```

File: plugins/pipelines/update.js [hunk 4]
```diff
@@ -136,6 +141,32 @@ module.exports = () => ({
...
+                        checkoutUrl: formattedCheckoutUrl,
+                        token
+                    });
+                    const privateDeployKeyB64 = Buffer.from(privateDeployKey).toString('base64');
+
+                    await secretFactory.create({
```
Comment (tkyi) [inline plugins/pipelines/update.js:161]: should this have a try/catch with error logs?

Comment (pritamstyz4ever) [reply plugins/pipelines/update.js:161]: seems this doesn't throw an error, we are anyway checking if the secret already exists before creating, also API call  should be logged already

**Review by tkyi — APPROVED** [review]

File: plugins/pipelines/create.js [hunk 5]
```diff
@@ -89,7 +90,10 @@ module.exports = () => ({
 
                 await defaultCollection.update();
             }
-            if (autoKeysGeneration) {
+            // check if pipeline has deploy key annotation then create secrets
+            const deployKeyAnnotation = pipeline.annotations && pipeline.annotations[ANNOTATION_USE_DEPLOY_KEY]
+      
+            if (autoKeysGeneration || deployKeyAnnotation) {
```
[the remark this answers is not in the document]
Comment (pritamstyz4ever) [reply plugins/pipelines/create.js:96]: yes, similar to chainPR. here : https://github.com/screwdriver-cd/models/blob/master/lib/pipeline.js#L876-L878
